File: slices/apps/apps-core/src/bridge.ts
import { Schema } from 'effect'
import { Bridge } from 'effect-messaging-core'

/**
 * Web → Host: the apps SPA asks the host to start a tunnel so an app
 * that requires a publicly-reachable origin can be launched. The host
 * either opens a tunnel and replies with {@link TunnelStarted} carrying
 * the new origin, or replies with {@link TunnelFailed} on error.
 */
const RequestTunnel = Schema.parseJson(Schema.TaggedStruct('RequestTunnel', {}))

/**
 * Web → Host: the apps SPA asks the host to open a launched app in a separate,
 * less-privileged native webview popup (presented by the Tauri host via
 * `tauri-plugin-native-webview`, with native chrome) instead of navigating the
 * main webview away from the SPA. `url` is the fully-resolved launch URL (origin
 * + `/apps/{id}`); the host opens it, following the launch redirect itself.
 * Fire-and-forget — there is no host→web reply, and a non-Tauri (standalone-web)
 * page never emits this (it navigates directly instead).
 */
const RequestSandboxedWebView = Schema.parseJson(
  Schema.TaggedStruct('RequestSandboxedWebView', { url: Schema.String })
)

/**
 * Host → Web: the host successfully started a tunnel; `origin` is the
 * publicly-reachable origin the SPA should redirect to in order to
 * launch a tunnel-requiring app.
 */
const TunnelStarted = Schema.parseJson(
  Schema.TaggedStruct('TunnelStarted', { origin: Schema.String })
)

/**
 * Host → Web: the host failed to start a tunnel; `reason` is a short
 * human-readable description suitable for surfacing as an inline error.
 */
const TunnelFailed = Schema.parseJson(
  Schema.TaggedStruct('TunnelFailed', { reason: Schema.String })
)

type AppsBridge = Bridge.Bridge<
  'Apps',
  {
    TunnelStarted: typeof TunnelStarted
    TunnelFailed: typeof TunnelFailed
  },
  {
    RequestTunnel: typeof RequestTunnel
    RequestSandboxedWebView: typeof RequestSandboxedWebView
  }
>

/**
 * Slice-level bridge between the embedded apps SPA and the Tauri/Expo host.
 * Web→Host carries the tunnel-request control signal and the
 * open-in-sandboxed-webview launch request; Host→Web carries the tunnel
 * start/fail outcome the SPA waits on before redirecting to a tunnel-requiring
 * app. The transport-level `__Ready` handshake already covers mount
 * synchronisation — no slice-level `Ready` tag.
 */
const AppsBridge: AppsBridge = Bridge.make({
  name: 'Apps',
  hostToWeb: [
    ['TunnelStarted', TunnelStarted],
    ['TunnelFailed', TunnelFailed],
  ] as const,
  webToHost: [
    ['RequestTunnel', RequestTunnel],
    ['RequestSandboxedWebView', RequestSandboxedWebView],
  ] as const,
})

export { AppsBridge, RequestSandboxedWebView, RequestTunnel, TunnelFailed, TunnelStarted }

File: slices/apps/apps-rust/src/bridge.rs
//! The apps bridge wire types shared with the embedded SPA.
//!
//! The apps SPA asks the host to bring a tunnel up (so an app that needs a
//! publicly-reachable origin can launch) via the web→host `RequestTunnel`
//! message; the host replies with [`AppsHostToWeb::TunnelStarted`] carrying the
//! verified public origin, or [`AppsHostToWeb::TunnelFailed`] with a reason.
//!
//! Payload shapes are pinned by the TS schema in
//! `slices/apps/apps-core/src/bridge.ts` — the web side validates inbound
//! payloads against that schema, so the serde representation here must stay
//! byte-compatible. The inbound `RequestTunnel` envelope (`{"_tag":
//! "RequestTunnel"}`) carries no fields, so the host decodes it by tag alone.

use serde::{Deserialize, Serialize};

/// Host→web messages on the apps bridge — the outcome the SPA waits on before
/// redirecting to a tunnel-requiring app.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "_tag")]
pub enum AppsHostToWeb {
    /// The host brought the tunnel up; `origin` is the verified public origin
    /// (`https://{host}`) the SPA should redirect to.
    /// Wire: `{"_tag":"TunnelStarted","origin":"https://host"}`.
    TunnelStarted { origin: String },
    /// The host could not bring the tunnel up; `reason` is a short
    /// human-readable description suitable for surfacing inline.
    /// Wire: `{"_tag":"TunnelFailed","reason":"..."}`.
    TunnelFailed { reason: String },
}

/// The web→host `RequestTunnel` tag literal, pinned against the TS schema's
/// `Schema.TaggedStruct('RequestTunnel', {})` in `apps-core/src/bridge.ts`.
pub const REQUEST_TUNNEL_TAG: &str = "RequestTunnel";

/// The web→host `RequestSandboxedWebView` tag literal, pinned against the TS
/// schema's `Schema.TaggedStruct('RequestSandboxedWebView', { url })` in
/// `apps-core/src/bridge.ts`. The host (wildflower-tauri) decodes the `url`
/// field and opens it in the shared sandboxed webview.
pub const REQUEST_SANDBOXED_WEBVIEW_TAG: &str = "RequestSandboxedWebView";

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tunnel_started_serializes_to_the_pinned_wire_format() {
        let message = AppsHostToWeb::TunnelStarted {
            origin: "https://dev1.example.com".to_string(),
        };
        assert_eq!(
            serde_json::to_string(&message).expect("serialize"),
            r#"{"_tag":"TunnelStarted","origin":"https://dev1.example.com"}"#
        );
    }

    #[test]
    fn tunnel_failed_serializes_to_the_pinned_wire_format() {
        let message = AppsHostToWeb::TunnelFailed {
            reason: "tunnel not available".to_string(),
        };
        assert_eq!(
            serde_json::to_string(&message).expect("serialize"),
            r#"{"_tag":"TunnelFailed","reason":"tunnel not available"}"#
        );
    }

    #[test]
    fn host_to_web_round_trips() {
        for message in [
            AppsHostToWeb::TunnelStarted {
                origin: "https://h".to_string(),
            },
            AppsHostToWeb::TunnelFailed {
                reason: "nope".to_string(),
            },
        ] {
            let encoded = serde_json::to_string(&message).expect("serialize");
            let decoded: AppsHostToWeb = serde_json::from_str(&encoded).expect("deserialize");
            assert_eq!(decoded, message);
        }
    }

    #[test]
    fn request_tunnel_tag_matches_the_ts_schema() {
        // The inbound envelope is `{"_tag":"RequestTunnel"}` (no fields); the
        // host routes on this literal.
        assert_eq!(REQUEST_TUNNEL_TAG, "RequestTunnel");
    }

    #[test]
    fn request_sandboxed_webview_tag_matches_the_ts_schema() {
        // The inbound envelope is
        // `{"_tag":"RequestSandboxedWebView","url":"..."}`; the host routes on
        // this literal and reads the `url` field.
        assert_eq!(REQUEST_SANDBOXED_WEBVIEW_TAG, "RequestSandboxedWebView");
    }
}
